main/src/ca/uwaterloo/flix/language/phase/TypeReconstruction.scala

```diff
@@ -0,0 +1,997 @@
+        case sym: Symbol.RestrictableCaseSym => throw InternalCompilerException(s"unexpected symbol: $sym", sym.loc)
+        case sym: Symbol.VarSym => throw InternalCompilerException(s"unexpected symbol: $sym", sym.loc)
+        case sym: Symbol.KindedTypeVarSym => throw InternalCompilerException(s"unexpected symbol: $sym", sym.loc)
+        case sym: Symbol.UnkindedTypeVarSym => throw InternalCompilerException(s"unexpected symbol: $sym", sym.loc)
+        case sym: Symbol.LabelSym => throw InternalCompilerException(s"unexpected symbol: $sym", SourceLocation.Unknown)
+        case sym: Symbol.HoleSym => throw InternalCompilerException(s"unexpected symbol: $sym", sym.loc)
+      }
+
+      groups.map {
+        case (k, v) => (k, v.toList)
+      }
+
+  }
+
+  /**
+    * Reconstructs types in the given defs.
+    */
+  private def visitDefs(root: KindedAst.Root, substs: Map[Symbol.DefnSym, Substitution], oldRoot: TypedAst.Root, changeSet: ChangeSet)(implicit flix: Flix): Map[Symbol.DefnSym, TypedAst.Def] = {
+    flix.subphase("Defs") {
+      val (staleDefs, freshDefs) = changeSet.partition(root.defs, oldRoot.defs)
+      ParOps.mapValues(staleDefs) {
+        case defn => visitDef(defn, root, substs(defn.sym))
+      } ++ freshDefs
+    }
+  }
+
+  /**
+    * Reconstructs types in the given classes.
+    */
+  private def visitClasses(root: KindedAst.Root, defSubsts: Map[Symbol.DefnSym, Substitution], sigSubsts: Map[Symbol.SigSym, Substitution], oldRoot: TypedAst.Root, changeSet: ChangeSet)(implicit flix: Flix): Map[Symbol.ClassSym, TypedAst.Class] = {
+    flix.subphase("Classes") {
+      val (staleClasses, freshClasses) = changeSet.partition(root.classes, oldRoot.classes)
+      ParOps.mapValues(staleClasses)(visitClass(_, root, defSubsts, sigSubsts)) ++ freshClasses
+    }
+  }
+
+  /**
+    * Reconstructs types in the given instances.
+    */
+  private def visitInstances(root: KindedAst.Root, substs: Map[Symbol.DefnSym, Substitution])(implicit flix: Flix): Map[Symbol.ClassSym, List[TypedAst.Instance]] = {
+    flix.subphase("Instances") {
+      ParOps.mapValues(root.instances) {
+        case insts => insts.map(visitInstance(_, root, substs))
+      }
+    }
+  }
+
+  /**
+    * Reconstructs types in the given class.
+    */
+  private def visitClass(clazz: KindedAst.Class, root: KindedAst.Root, defSubsts: Map[Symbol.DefnSym, Substitution], sigSubsts: Map[Symbol.SigSym, Substitution]): TypedAst.Class = clazz match {
+    case KindedAst.Class(doc, ann, mod, sym, tparam0, superClasses0, assocs0, sigs0, laws0, loc) =>
+      val tparam = visitTypeParam(tparam0, root) // TODO ASSOC-TYPES redundant?
+      val superClasses = superClasses0 // no subst to be done
+      val assocs = assocs0.map {
+        case KindedAst.AssocTypeSig(doc, mod, sym, tp0, kind, loc) =>
+          val tp = visitTypeParam(tp0, root) // TODO ASSOC-TYPES redundant?
+          TypedAst.AssocTypeSig(doc, mod, sym, tp, kind, loc) // TODO ASSOC-TYPES trivial
```

Comment:
What about these comments? Do you want to remove them after everything is in place?

Reply:
I'd like to wait on these. I'm pretty sure this can be optimized but I prefer it to be correct first :)